First pass at a hooks definition. (#45)
open-feature/spec: specification/flag-evaluation/hooks.md

## Overview

> Hooks are a mechanism whereby application developers can add arbitrary behavior

Reviewer: This is a good opportunity for us to use the User Roles in the glossary. Maybe we could add something like:

They operate similarly to middleware in many web frameworks. Integration Authors create integrations between external systems and the OpenFeature flag evaluation workflow in the form of hooks, which an Application Integrator can then enable within their Feature Flag SDK configuration.

to flag evaluation. They operate similarly to middleware in many web frameworks.

### Definitions

**Stage**: An explicit portion of the flag evaluation lifecycle. e.g. `before` being "before the evaluation is run.
**Invocation**: A single call to evaluate a flag. `client.getBooleanValue(..)` is an invocation.

> **API**: The global API singleton.

Reviewer: Could we add a definition for hook itself here?

Hook: a function (callback?) that is called by the OpenFeature framework at a specific invocation stage, with hook context passed in.

### Hook context

Hook context exists to provide hooks with information about the invocation.

> Hook context **MUST** provide: the flag key, evaluation context, default value and a list of  executed hooks by stage.

> > Condition: You **MUST** provide `flag type` if the language type system differentiates between strings, numbers, booleans, and structures.

Reviewer: not sure this needs to be conditional. Given that the evaluation and provider API requires type-specific methods, wouldn't we always know the flag type regardless of the host language's type system?

> > Hook context **SHOULD** provide: provider, client

Reviewer: are provider and client an actual runtime instance of the provider and client that the hook could interact with? or just a name? or something else?

Reviewer 2: In the demo repo, it's the instances. We have some hooks (OTel for instance) that grabs some metadata from them (such as the provider name, client version, client name, etc). This is a good point.

> > flag key, flag type, default value properties **MUST** be immutable.

Reviewer: maybe add something here like "if the language does not support immutability, the hook MUST NOT modify these properties"

> The evaluation context **MUST** be mutable only within the `before` hook.

### State

> State **MUST** be a map of objects.

> Condition: State **MUST** be immutable, if the implementation language supports a mechanism for marking data as immutable.

### Hook creation and parameters

> Hooks **MUST** specify at least one stage.

> > The `before` stage *MUST* run before flag evaluation occurs. It accepts a `hook context` (required) and `state` (optional) as parameters and returns either a `HookContext` or nothing.

Reviewer: given the definitions above, I think it would be more precise to say "*MUST* run before invocation occurs"

Author: That seems not quite right to me. The application author writes something like:

```
client.getBooleanValue(...)
```

That's a flag invocation.

Within the framework code, it does something like:

```
runBefore(hooks)
callProviderToGetRealData()
runAfter(hooks)
```

This is trying to say that the `runBefore` function needs to happen before the `callProviderToGetRealData` function.

Reviewer: ah got it. I misunderstood the meaning of invocation in the definitions. I thought it just meant a specific flag evaluation, but I think the intention is for invocation to mean the call that the Application Author makes into the evaluation API, vs the call the framework makes into the provider API.

```
HookContext|void before(HookContext, State)
```

> > The `after` stage **MUST** run after flag evaluation occurs. It accepts a `hook context` (required), `flag evaluation details` (required) and `state` (optional). It has no return value.

Reviewer 2: We could say "flag evaluation invocation".

> > The `after` stage **MUST** run after flag evaluation occurs. It accepts a `hook context` (required), `flag evaluation details` (required) and `state` (optional). It has no return value.

Reviewer: no I think I just misunderstood what invocation was defining. We can leave it as is.

> > The `error` hook runs when errors are encountered in the `before` or `after` state or when flag evaluation errors. It accepts `hook context` (required), `exception` for what went wrong (required), and `state` (optional). It has no return value.

Reviewer: s/state/stage/

> The `finally` hook **MUST** run after the `before`, `after`, and `error` stages. It accepts a `hook context` (required) and `state` (optional). There is no return value.

> > Condition: If `finally` is a reserved word in the language, `afterAll` should be used.

Reviewer: afterAll might be a confusing name for people who associate that with testing frameworks. In testing frameworks it has a different semantic meaning - run this hook once after all the tests have finished.

would `eventually` be better? or `alwaysAfter`? or `finallyAfter`?

Author: Going w/ `finallyAfter`.

### Hook registration & ordering

> The API, Client and invocation **MUST** have a method for registering hooks which accepts `flag evaluation options`

```js
OpenFeature.addHooks(new Hook1());

//...

Client client = OpenFeature.getClient();
client.addHooks(new Hook2());

//...

client.getValue('my-flag', 'defaultValue', new Hook3());
```

> > Hooks **MUST** be evaluated in the following order: API before, client before, invocation before, invocation after, client after, API after, invocation finally, client finally, API finally. If errors are present, the order is: invocation error, client error, API error.

Reviewer: for the errors scenario, I think it would be clearly to specify the full sequence, which I _think_ is:

- API before
- client before
- invocation before
- invocation error
- client error
- API error
- invocation finally
- client finally
- API finally

```
// here we are using the order defined above (note if it wasn't defined, we'd have to order them somehow anyway
Hooks[] allHooks = [ ...OpenFeature.getHooks(), ...this.getHooks(), ...evaluationOptions.hooks ]

try {
  // evaluation all our before hooks
  for (h in allHooks) {
    h.before(...);
  }

  provider.resolveValue(...);

  // evaluation all our after hooks
  for (h in allHooks.reverse()) {
    h.after(...);
  }
} catch (Error err) {
  // evaluation all our error hooks
  for (h in allHooks.reverse()) {
    h.error(...);
  }
} finally {
  // evaluation all our finally hooks
  for (h in allHooks.reverse()) {
    h.finally(...);
  }
}
```

> > When an error occurs in hook evaluation, further evaluation **MUST** stop and error hook evaluation begins.

Reviewer: in this section I was a bit confused about when we were talking about errors in the provider evaluation vs errors in the hook evaluation. It might be clearer if the language was a bit more specific. For example, rather than saying "if an error is encountered in the error stage" we could say "if an error is encountered during evaluation of a hook in the error stage", or just "during evaluation of an error hook"

Author: > If an error occurs during the evaluation of `before` or `after` hooks, any remaining hooks in the `before` or `after` stages **MUST** not be invoked.

^ I'm going to try that. The intent is errors are kinda like an exception in that it stops the flow and bubbles up to the error stage.

> > If an error is encountered in the error stage, it **MUST** not be returned to the user.

Reviewer: for consistency, it seems like this should also be the case for errors in the finally stage?

Author: The intent of this line was:  If you have an error from the provider and that somehow triggers an error in the error flow.. don't overwrite their original error with this new one.

I'm not actually clear what an error in finally should do. Our whole MO seems to be "never throw, except maybe around client initialization", but also I don't want to have this logic complexity in the finally handler.

```python
try:
  before()
  eval()
  after()
except:
  error()
finally:
  try:
    finally()
  except:
    error()
```

> If an error occurs in the `before` or `after` hooks, the `error` hooks **MUST** be invoked.

> If an error occurs in the `finally` hook, it **MUST** not trigger the `error` hook.

### [Flag evaluation options](../types.md#evaluation-options)

> `Flag evalution options` must contain a list of hooks to evaluate.

> > `Flag evaluation options` may contain `state`, a map of data to be provided to hook invocations.

Reviewer: maybe we'd specify this somewhere else, but I'm left a bit unclear as to how exactly an Application Author would pass this state in to an invocation

Reviewer 2: I imagine this would be done via the [EvaluationOptions](https://github.com/open-feature/spec/blob/main/specification/types.md#evaluation-options). To your earlier point, I think a `HookHints` option would be apropos here.

## Overview

Hooks are a mechanism whereby application developers can add arbitrary behavior
to flag evaluation. They operate similarly to middleware in many web frameworks.

### Definitions

**Hook**: Application author/integrator-supplied logic that is called by the OpenFeature framework at a specific stage.
**Stage**: An explicit portion of the flag evaluation lifecycle. e.g. `before` being "before the evaluation is run.
**Invocation**: A single call to evaluate a flag. `client.getBooleanValue(..)` is an invocation.
**API**: The global API singleton.

### Hook context

Hook context exists to provide hooks with information about the invocation.

###### Requirement 1.1

> > Hook context **MUST** provide: the flag key, evaluation context, default value and a list of  executed hooks by stage.

Reviewer 2: Did we decide to remove the "list of executed hooks by stage"?

##### Condition 1.2

> If the language type system differentiates between strings, numbers, booleans, and structures.
>
>##### Condition 1.2.1
>

> > Condition: You **MUST** provide `flag type`.

Reviewer 2: I think we decide to remove this condition, and just make this a requirement.

Also, avoid use of "you". Use imperative mood / passive voice:

"`Flag type` must be provided" or "`Hook context must contain the flag type`"

##### Requirement 1.2

> Hook context **SHOULD** provide: provider, client

##### Requirement 1.3

> flag key, flag type, default value properties **MUST** be immutable. If the language does not support immutability, the hook **MUST NOT** modify these properties.

##### Requirement 1.4

> The evaluation context **MUST** be mutable only within the `before` hook.

### HookHints

##### Requirement 2.1

> HookHints **MUST** be a map of objects.

##### Condition 2.2

> The implementation language supports a mechanism for marking data as immutable.
>
> ##### 2.2.1
>
> Condition: HookHints **MUST** be immutable.

### Hook creation and parameters

##### Requirement 3.1

> Hooks **MUST** specify at least one stage.

##### Requirement 3.2

> > The `before` stage **MUST** run before flag evaluation occurs. It accepts a `hook context` (required) and `state` (optional) as parameters and returns either a `HookContext` or nothing.

Reviewer 2: Seems like we have a conflict here between `state` and `HookHints`? They are the same thing, right?

## Overview

Hooks are a mechanism which can tie into the lifecycle of flag evaluation. They
operate similarly to middleware in many web frameworks.

> 

Reviewer 2: "tie in" is vague. I suggest something like "hooks are a mechanism by which arbitrary behavior (or maybe "logic" if you prefer) to flag evaluation".

### Hook context

Hook context exists to provide hooks with information about the invocation.

> > Hook context **MUST** provide: the flag key, flag type, evaluation context, default value and a list of  executed hooks by stage.

Reviewer 2: > flag type may not be possible in some languages. You may want to make this conditional, like some of the other sections: 

> The language type system differentiates between strings, numbers, booleans and structures.

Author: We would still know what type was requested, right?  Like they will have to call `getBooleanValue`, no?

Reviewer 2: > We would still know what type was requested, right? Like they will have to call getBooleanValue, no?

"getBooleanValue" only exists in languages implementations that have types. A completely type-less language would just implement `getValue`

Author: Updated to be conditional.

Reviewer: I'm still confused as to how we wouldn't have that information, even in a completely type-less language. Why would a type-less language just implement getValue? Wouldn't we still track the type as metadata somewhere? Probably a conversation to be had outside of this PR.

Reviewer 2: There exists languages where there simply isn't rigorous concepts around data, even at runtime. There are not types - but just "bytes" that can be interpreted in multiple ways.

We will probably never deal with such languages, so I'm not married to keeping this requirement.

> Hook context **SHOULD** provide: provider, client

> flag key, flag type, default value properties **MUST** be immutable.

> > The evaluation context **MUST** be mutable only within the `before` hook.

Reviewer 2: I like both of these ideas, but I'm not sure all languages have immutability concepts, so you might want to make this conditional to available language mechanisms.

### Hook creation and parameters

> > Hooks **MUST** specify at least one stage: `before(HookContext, ImmutableState)`, `after(HookContext, FlagEvaluationDetails, , ImmutableState)`, `error(HookContext, Exception, ImmutableState)`, or `finally(HookContext, ImmutableState)`.

Reviewer 2: ```suggestion
> Hooks **MUST** specify at least one stage: `HookContext | void before(HookContext, ImmutableState)`, ` FlagEvaluationDetails | void after(HookContext, FlagEvaluationDetails, , ImmutableState)`, `void error(HookContext, Exception, ImmutableState)`, or `void finally(HookContext, ImmutableState)`.
```

Reviewer 2: I added the return values (which I think are fairly important).

I wonder if it's better to use plan prose instead of pseudo-code to describe the parameters and return values.

Reviewer 2: > Would be interested to hear what FlagEvaluationDetails return value is used for.

The idea was you could use it for for a few interesting things in after. For example. maybe you want to turn some JSON flag value into a class instance - you could mutate the `.value` field to actually return an instance - the hook would be a little factory that calls a constructor.

Author: @toddbaert This would imply that subsequent hooks couldn't assume that the flag value is the type that the flag says it is, right?

Reviewer 2: No, but we might support an generic arg in order to provide that type information in a hook definition. I'm not 100% convinced we need this feature, but it seems like it could be useful.

Author: https://github.com/open-feature/spec/issues/48

> > Condition: If `finally` is a reserved word in the language, `afterAll` should be used.

Reviewer 2: "stage" is a good term for this.

### Hook registration & ordering

> > The API, Client and invocation **MUST** have a method for registering hooks which accepts `flag evaluation options`

Reviewer 2: is "API" or "global" a better descriptor here?

Reviewer 2: That works for me.

> > Hooks **MUST** be evaluated in the following order: API before, client before, invocation before, invocation after, client after, API after, invocation finally, client finally, API finally. If errors are present, the order is: invocation error, client error, API error.

Reviewer 2: We've used "invocation" here a few times, and I think it may be good to explain that "invocation" here is specifically a "flag evaluation invocation". Maybe when you format this, you could do this in a supplemental bit of text, not in a normative spec requirement.

> When an error occurs in hook evaluation, further evaluation **MUST** stop and error hook evaluation begins.

> > If an error is encountered in the error flow, it **MUST** not be returned to the user.

Reviewer 2: Throwing from an error stage is something we said could be a possible exception to the "never throw" rule. If we want this, we can make such an exception in the `flag evaluation` spec. I think it has some utility.

@beeme1mr @agentgonzo @moredip what do you guys think?

Reviewer 3: If we're going to have a rule of "never throw", then I don't think we should start adding exceptions to it.

IMO, a hook is a non-critical part of the evaluation flow (the critical part being the flag evaluation/resolution itself). If you throw an error, it means that the `getXyzValue()` won't return an evaluated flag value to the user (bad) IIUI.

Reviewer 3: Is it important to actually return an error in any of the hooks to the caller? What would they do with such an error? Would it be sufficient to merely log the error in the provider?

> If an error occurs in the `before` or `after` hooks, the `error` hooks **MUST** be invoked.

> If an error occurs in the `finally` hook, it **MUST** not trigger the `error` hook.

> ### Flag evaluation options

Reviewer 2: You may want to link to the `types.md` file here: https://github.com/open-feature/spec/blob/062a083702a442aee184e2838c90911eb8b778b5/specification/types.md

> `Flag evalution options` must contain a list of hooks to evaluate.

> > `Flag evaluation options` may contain `immutable state`, a map of data to be provided to hook invocations.

Reviewer 2: Though the state is immutable, I think I'd prefer not to see that detail be part of the interfaces, especially since not all languages support immutable structures. I think `state` is fine, I've seen such things in other APIs.

> > Immutable hook data **MUST** be passed to each hook through a parameter. It is merged into the object in the precedence order API -> Client -> Invocation.

Reviewer 2: ```suggestion
> The `immutable state` **MUST** be passed to each hook through a parameter. It is merged into the object in the precedence order API -> Client -> Invocation.
```

> > Immutable hook data **MUST** be passed to each hook through a parameter. It is merged into the object in the precedence order API -> Client -> Invocation.

Author: Updated it to be last wins and included a code sample.

> > The hook **MUST** not alter the `immutable state` object.

Reviewer 2: The way this is phrased seems like it targets the hook implementer, not the SDK implementer. I think it should instead say that the `immutable state` should ideally be an immutable object, if the language has such concepts (perhaps use a condition).

### Hook evaluation

> `Immutable state` **MUST** passed between each hook.

> > As hooks are run, they **MUST** be added to an `evaluatedHooks` property within the `EvaluationContext`.

Reviewer 4: There was some debate on whether or not this was a good idea to specify in the spec.

https://github.com/open-feature/spec/pull/35#discussion_r855316168

We concluded that it's a good idea but not necessary at this stage of the spec. What do you think?

Author: I'm game to drop it for now and see if the need pops up before GA.

> > If a hook throws an error, it **MUST** be in `evaluatedHooks`.

Reviewer 4: This requirement isn't clear to me. Could you elaborate what you mean?

Author: Add the current hook to the evaluated hooks before running it so, if it throws an error, you see it in evaluated hooks. 

Do:
```
Evaluated hooks.add(h)
h.run()
```

Not:
```
H.run()
Evaluated hooks.add(H)
```

Does that make sense?

Reviewer 2: I think this is a bit redundant, it's implicit in the original statement. But I agree we don't need it now.

## Overview

Hooks are a mechanism whereby application developers can add arbitrary behavior
to flag evaluation. They operate similarly to middleware in many web frameworks.

### Definitions

**Stage**: An explicit portion of the flag evaluation lifecycle. e.g. `before` being "before the evaluation is run.
**Invocation**: A single call to evaluate a flag. `client.getBooleanValue(..)` is an invocation.
**API**: The global API singleton.

### Hook context

Hook context exists to provide hooks with information about the invocation.

> Hook context **MUST** provide: the flag key, flag type, evaluation context, default value and a list of  executed hooks by stage.

> Hook context **SHOULD** provide: provider, client

> flag key, flag type, default value properties **MUST** be immutable.

> The evaluation context **MUST** be mutable only within the `before` hook.

> ### State

Reviewer 5: What is the exact use case of `State`? One question I have is does it risk causing flags to evaluate differently on different instances of a service?

Author: I'm primarily interested in, on a per-invocation basis, specifying information to a hook. Imagine I have a hook that emits metrics. I want that in almost all cases. I'd like the ability to say "For this particular request.. don't emit metrics".   I'm thinking we use `state` for that.

Reviewer 5: That makes sense, I'm wondering if there's a way we can clarify or scope this expected usage down. We're quickly layering on top of the evaluation process, hooks, then state without an initial release.

Reviewer 2: I'm OK with the concept, and I think I see the value from other discussions with @justinabrahms . It might be that we need an example use-case or other form of explanation (and perhaps to change the name to something that more clearly defines the purpose; "state" may be a bit vague?).

Once this is formatted to a spec document, that supplementary info (and even a code sample, perhaps) can be added in a non-normative section.

Author: @beeme1mr I have a very concrete use-case for it. I'd _prefer_ to keep it in this spec, as when it merges I'm going to update the java sdk code to do the thing I'm saying and then I can use the code internally. That said, if you feel strongly, I'll move it out.

The explicit use-case here is I have a metrics service. Sometimes, we want to control precisely when metrics are emitted (due to some pre-fetching logic we have), but usually not. I want to register a "please emit the relevant metrics" hook which triggers by default.

Sometimes, when I want to control when those metrics emit, I want to not have that piece of code run. To do that, I have to provide some state to the function. The options for state that I've considered:

1. We could use `evaluatedHooks`. I could register a `DoNotEmitHooks` hook, which is a sentinel value that `EmitHooks` looks for. If it sees it in the evaluation order before it.. it won't emit hooks. (eew).
2. We could use `EvaluationContext` which is the map of data we send up for evaluation. It could have users put data in there to communicate to the hooks, but isn't really used for targeting. (eew)
3. We could pass some explicit hook-specific state. That's what I'm proposing here.
4. We could call methods on the `EmitMetricsHook` itself like `dontEmitMetricsOnNextCall`, but I worry about the thread safety of this approach.
5. Maybe a secret 5th option that you can think of?

@moredip I hear you about rapidly iterating. This came up when we were trying to port one of the hooks we care about into the janky java sdk I wrote. We didn't actually have the relevant primitives to accomplish the goal, hence the proposal.

Reviewer: I found the work "state" a bit misleading. I assumed it was a place for hooks to maintain state that could then be picked up by the hook in a later evaluation stage. I think what's being describe here is more like per-invocation configuration for a hook. Assuming I'm reading that correctly, then maybe "Hook Hints" would be a clearer name than state?

Reviewer 2: @moredip  Agreed, that's really my primary concern with this requirement as well.

Author: Moved to `HookHints`. Much better name. :) Can we resolve this thread?

Reviewer 5: 👍 on resolving

> State **MUST** be a map of objects.

> > Condition: If your language supports it, state **MUST** be immutable.

Reviewer 4: ```suggestion
> Condition: State **MUST** be immutable, if the implementation language supports a mechanism for marking data as immutable.
```

### Hook creation and parameters

> Hooks **MUST** specify at least one stage.

> The `before` stage *MUST* run before flag evaluation occurs. It accepts a `hook context` (required) and `state` (optional) as parameters and returns either a `HookContext` or nothing.

```
HookContext|void before(HookContext, State)
```

> The `after` stage **MUST** run after flag evaluation occurs. It accepts a `hook context` (required), `flag evaluation details` (required) and `state` (optional). It has no return value.

> The `error` hook runs when errors are encountered in the `before` or `after` state or when flag evaluation errors. It accepts `hook context` (required), `exception` for what went wrong (required), and `state` (optional). It has no return value.

> The `finally` hook **MUST** run after the `before`, `after`, and `error` stages. It accepts a `hook context` (required) and `state` (optional). There is no return value.

> Condition: If `finally` is a reserved word in the language, `afterAll` should be used.

### Hook registration & ordering

> The API, Client and invocation **MUST** have a method for registering hooks which accepts `flag evaluation options`

> > Hooks **MUST** be evaluated in the following order: API before, client before, invocation before, invocation after, client after, API after, invocation finally, client finally, API finally. If errors are present, the order is: invocation error, client error, API error.

Reviewer 5: To me on initial pass this appears to be a lot of hook stages. Have we actually verified that there is use cases for each one? In my view it'll be cleaner to start with few hook stages and add them as needed vs having a large number upfront and finding it's becoming a maintenance or implementation headache.

Reviewer 2: I see your concern. It think it looks like there's a lot more complexity here than there actually is. The There is really only: `before`, `after`, `error`, and `finally` - it's just that there's 3 places to attach hooks, and so there's a need to define how they interact.

In code, there's not too much complexity here (I hope). From the Application Author's perspective, it's just:

```
OpenFeature.addHooks(new Hook1);

//...

Client client = OpenFeature.getClient();
client.addHooks(new Hook2);

//...

client.getValue('my-flag', 'defaultValue', new Hook3);
```

We are defining the order of these hooks relative to eachother, which isn't too tough in implementation logic either. It looks something like this:

```
// client logic...

// here we are using the correct order (note if it wasn't defined, we'd have to order them somewhow anyway
Hooks[] allHooks = [ ...OpenFeature.getHooks(), ...this.getHooks, ...evaluationOptions.hooks ]

// evaluation all our before hooks
for (h in allHooks) {
  h.before();
}

provider.resolveValue(...);

// evaluation all our after hooks
for (h in allHooks.reverse()) {
  h.after()
}

// more client logic...
```

Let me know if that doesn't address your concern. It might be that there's an easier way to describe this, maybe with a point for list, or with prose. I'm not sure, but I get the point @justinabrahms was trying to make.

Reviewer 5: @toddbaert thanks for the example in code. That does clear it up.

Reviewer 2: You're welcome! If anything, this might mean that when we do the full markdown spec, adding code examples in non-normative auxiliary sections for this requirement is a good idea.
